src/ast/passes/codegen_llvm.cpp

```diff
@@ -719,37 +720,43 @@ void CodegenLLVM::visit(Call &call)
                                                              // probe_read_str's
                                                              // null byte
 
-      // largest read we'll allow = our global string buffer size
-      Value *max = b_.getInt64(bpftrace_.config_.get(ConfigKeyInt::max_strlen));
       // integer comparison: unsigned less-than-or-equal-to
       CmpInst::Predicate P = CmpInst::ICMP_ULE;
       // check whether proposed_strlen is less-than-or-equal-to maximum
-      Value *Cmp = b_.CreateICmp(P, proposed_strlen, max, "str.min.cmp");
+      Value *Cmp = b_.CreateICmp(P, proposed_strlen, strlen, "str.min.cmp");
       // select proposed_strlen if it's sufficiently low, otherwise choose
       // maximum
-      Value *Select = b_.CreateSelect(
-          Cmp, proposed_strlen, max, "str.min.select");
-      b_.CreateStore(Select, strlen);
-    } else {
-      b_.CreateStore(
-          b_.getInt64(bpftrace_.config_.get(ConfigKeyInt::max_strlen)), strlen);
+      strlen = b_.CreateSelect(Cmp, proposed_strlen, strlen, "str.min.select");
     }
-    AllocaInst *buf = b_.CreateAllocaBPF(
-        bpftrace_.config_.get(ConfigKeyInt::max_strlen), "str");
-    b_.CREATE_MEMSET(
-        buf, b_.getInt8(0), bpftrace_.config_.get(ConfigKeyInt::max_strlen), 1);
+
+    Function *parent = b_.GetInsertBlock()->getParent();
+    BasicBlock *lookup_failure_block = BasicBlock::Create(
+        module_->getContext(), "scratch_lookup_failure", parent);
+    BasicBlock *lookup_merge_block = BasicBlock::Create(module_->getContext(),
+                                                        "scratch_lookup_merge",
+                                                        parent);
+
+    Value *buf = b_.CreateGetStrScratchMap(str_id_,
+                                           lookup_failure_block,
+                                           call.loc);
+    b_.CREATE_MEMSET(buf, b_.getInt8(0), max_strlen, 1);
     auto arg0 = call.vargs->front();
     auto scoped_del = accept(call.vargs->front());
-    b_.CreateProbeReadStr(ctx_,
-                          buf,
-                          b_.CreateLoad(b_.getInt64Ty(), strlen),
-                          expr_,
-                          arg0->type.GetAS(),
-                          call.loc);
-    b_.CreateLifetimeEnd(strlen);
+    b_.CreateProbeReadStr(
+        ctx_, buf, strlen, expr_, arg0->type.GetAS(), call.loc);
+    b_.CreateBr(lookup_merge_block);
+
+    // Think of this like an assert(). In practice, we cannot fail to lookup a
+    // percpu array map unless we have a coding error. Rather than have some
```

Comment:
By "coding error" you mean we pass in a key that does not exist right?

Reply:
Yes